stdlib/http/src/main/ballerina/http/http_secure_client.bal

```diff
@@ -294,178 +345,603 @@ public function createHttpSecureClient(string url, ClientEndpointConfig config)
     }
 }
 
-# Prepare HTTP request with the required headers for authentication.
+# Prepare HTTP request with the required headers for authentication based on the scheme and return a flag saying whether
+# retry is required if the response will be 401.
 #
 # + req - An HTTP outbound request message
 # + config - Client endpoint configurations
-# + return - The Error occured during HTTP client invocation
-function generateSecureRequest(Request req, ClientEndpointConfig config) returns ()|error {
+# + tokenCache - Cached token configurations
+# + return - Whether retry is required for 401 response or `error` if error occured during HTTP client invocation
+function generateSecureRequest(Request req, ClientEndpointConfig config,
+                               CachedTokenConfig tokenCache) returns boolean|error {
     var auth = config.auth;
     if (auth is AuthConfig) {
         var authConfig = auth["config"];
         if (auth.scheme == BASIC_AUTH) {
             if (authConfig is BasicAuthConfig) {
-                string username = authConfig.username;
-                string password = authConfig.password;
-                string str = username + ":" + password;
-                string token = encoding:encodeBase64(str.toByteArray("UTF-8"));
-                req.setHeader(AUTH_HEADER, AUTH_SCHEME_BASIC + WHITE_SPACE + token);
+                string authToken = check getAuthTokenForBasicAuth(authConfig);
+                req.setHeader(AUTH_HEADER, AUTH_SCHEME_BASIC + WHITE_SPACE + authToken);
+                return false;
             } else {
-                error e = error(HTTP_ERROR_CODE, { message: "Basic auth config not provided" });
-                panic e;
+                string errMsg = "Basic auth config not provided.";
+                log:printError(errMsg);
```

Comment:
This error is returned to the user right? So do we still need an error log here? Applicable for other scenarios too.